client/task_runner_test.go

```diff
@@ -172,28 +168,22 @@ func TestTaskRunner_SaveRestoreState(t *testing.T) {
 
 	// Snapshot state
 	time.Sleep(200 * time.Millisecond)
-	err := tr.SaveState()
-	if err != nil {
+	if err := tr.SaveState(); err != nil {
 		t.Fatalf("err: %v", err)
 	}
 
 	// Create a new task runner
 	tr2 := NewTaskRunner(tr.logger, tr.config, upd.Update,
-		tr.ctx, tr.allocID, &structs.Task{Name: tr.task.Name})
-	err = tr2.RestoreState()
-	if err != nil {
+		tr.ctx, tr.allocID, &structs.Task{Name: tr.task.Name}, tr.restartTracker)
+	if err := tr2.RestoreState(); err != nil {
 		t.Fatalf("err: %v", err)
 	}
 	go tr2.Run()
 	defer tr2.Destroy()
 
 	// Destroy and wait
-	tr2.Destroy()
-
-	select {
-	case <-tr.WaitCh():
-	case <-time.After(2 * time.Second):
-		t.Fatalf("timeout")
+	time.Sleep(200 * time.Millisecond)
```

Comment:
Because of the goroutine on 181 I think this is actually a race condition if we don't use a channel or wait group. Maybe:

```
go func() {
    tr2.Run()
    close(doneCh)
}()
```